### ISSUE TITLE
atom2string: variants for displaying and saving atoms

### ISSUE THREAD

**Issue by umlaeute** [pure-data/pure-data#1310]
currently there is a single `atom_string` function, that is used to convert atoms into a string-representation.
this string representation is then used both to store atom values on disk, and to display them in the GUI.

having a uniform function is a nice idea, but unfortunately also has drawbacks:
- numbers:
  - for *displaying* numbers we want to have "nice looking" numbers (hopefully eating away any floating point inaccuracies). e.g. `440` cannot be properly represented as (double precision) number, but we don't want to see the more correct `439.99999243520523`.
  - for *saving* numbers, we do want to have the full resolution. E.g. both `10000000` and `10000001` (up to `10000050`) are currently saved as `1e07`. When restoring the atom from the string, these numbers all become `10000000` :-)
- symbols
  - when saving to disk, spaces in symbols need to be escaped with `\` (so they can be distinguished from lists of symbols)
  - when displaying symbols with spaces, we don't want to see those escaping backslashes (see https://github.com/pure-data/pure-data/issues/824)

**Comment by umlaeute** [comment]
oh, and (en lieu with #824), i think the "nice" values should displayed in:
- gatoms (numberbox, symbolbox)
- the Pd-console

whereas the "hi-res" values should be *displayed* in
- messageboxes (`[set $1( -> [$1(`)

ideally, objects and messageboxes should store (and display) the *literal* value that was entered, even if there are shorter representations without loosing precision (e.g. `400.00`)

i'm not sure about copy-paste.

**Comment by Spacechild1** [comment]
I agree with everything except

> whereas the "hi-res" values should be displayed in
>
>    * messageboxes ([set $1( -> [$1()

Having `[439.99999243520523(` instead of `[440(` after saving/reopening would be rather irritating... I think it's enough that the floats are stored with sufficient precision.

> ideally, objects and messageboxes should store (and display) the literal value that was entered,

That would be great, but I fear that it is not possible without some fundamental (breaking) changes...

**Comment by Spacechild1** [comment]
Here's an additional idea:

What about adding a flag to `[print]` to enable printing floats in full precision? With this we could always show the "short" output in all GUI elements, but the user still has the option to have a look at the "real" value via `[print]`, e.g. for debugging purposes.

**Comment by Spacechild1** [comment]
Also, we could add an option to `number2` to show floats in full precision.

**Comment by Spacechild1** [comment]
> but rather if you dynamically set a messagebox, you would see the data in a "lossless" representation.
that is:
>
> [440(
> |
> [set $1(
> |
> [439.99999243520523(

While it might be nice with cases like `[1000000.1( -> [set $1( -> [ (`, the example you gave would be rather confusing/irritating to most users.

I think a general guideline could be: don't show the "long" version unless explicitly requested.

**Comment by umlaeute** [comment]
so why do you say that an escaping character "must [...] be displayed to the programmer", but a number-representation an irritating confusion?

if messageboxes are code (rather than "user facing GUI objects like symbol atoms") wouldn't the appropriate behaviour be to represent the code?

it's not that i don't see the confusion (and i deliberately picked `440` as an example because it *is* irritating).
but i think it's also confusing if a symbol `foo bar` gets displayed as `foo bar` in a symbol box and in the (normal) `[print]`, but as `foo\ bar` in a message box.
(and it would be even more irritating if a messagebox that displays as  `[list foo bar(` would output a list with a single atom `foo bar`, whereas a messagebox that also displays as  `[list foo bar(` would output a list with two atoms `foo` and `bar`, because they are stored differently on filesystem.)

i would rather not introduce different representations of symbols vs floats on a case-by-case basis.

**Comment by umlaeute** [comment]
oh, and of course there's the case of `[symbol filename$1.wav(`, in which case i totally agree that the `$1` should expand to the nice-version (e.g. `440`)

**Comment by Spacechild1** [comment]
> so why do you say that an escaping character "must [...] be displayed to the programmer", but a number-representation an irritating confusion?

Good point. I would say that escaping characters in a string are important information in the sense that they *change* semantics, and I don't think that's true for different floating point number representations.

1) `[440(` vs `[439.99999243520523(`: both are obviously floating point numbers

2) `[list foo bar(` vs `[list foo\ bar(`: the first one is a two element list and second is a single symbol, which is a big difference and has to be communicated to the programmer.

But that's just my feeling/opinion...

> but i think it's also confusing if a symbol foo bar gets displayed as foo bar in a symbol box and in the (normal) [print], but as foo\ bar in a message box.
> (and it would be even more irritating if a messagebox that displays as [list foo bar( would output a list with a single atom foo bar, whereas a messagebox that also displays as [list foo bar( would output a list with two atoms foo and bar,

I think in the case of symbols we both agree that the second is much more confusing than the former. We just have to find a consensus on the topic of floating point numbers.


> oh, and of course there's the case of [symbol filename$1.wav(, in which case i totally agree that the $1 should expand to the nice-version (e.g. 440)

Good point!
